File: GDS_Unity/tools/render_osk.py
#!/usr/bin/env python3
"""Rasterize the primitives recorded by tools/render_osk.c with the real
atlas -> pixel-true preview of the actual C draw path.
Run after: gcc -O1 -o /tmp/render_osk tools/render_osk.c -lm &&
           (cd /tmp && /tmp/render_osk)"""
from PIL import Image
import os

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
AW = AH = 320
ATLAS = os.path.join(ROOT, 'ports/gamedevstory/gamedevstory/osk_font.rgba')
atlas = Image.frombytes('RGBA', (AW, AH), open(ATLAS, 'rb').read())

def render(prim_path, out_name):
    im = Image.new('RGBA', (640, 480), (90, 60, 130, 255))   # game-ish purple bg
    for line in open(prim_path):
        p = line.split()
        if not p or p[0] == 'END':
            continue
        if p[0] == 'RECT':
            x, y, w, h = (float(v) for v in p[1:5])
            r, g, b = (float(v) for v in p[5:8])
            tile = Image.new('RGBA', (max(1, round(w)), max(1, round(h))),
                             (int(r*255), int(g*255), int(b*255), 255))
            im.alpha_composite(tile, (round(x), round(y)))
        elif p[0] == 'RECTA':
            x, y, w, h = (float(v) for v in p[1:5])
            r, g, b, a = (float(v) for v in p[5:9])
            tile = Image.new('RGBA', (max(1, round(w)), max(1, round(h))),
                             (int(r*255), int(g*255), int(b*255), int(a*255)))
            im.alpha_composite(tile, (round(x), round(y)))
        elif p[0] == 'QUAD':
            x, y, w, h = (float(v) for v in p[1:5])
            u0, v0, u1, v1 = (float(v) for v in p[5:9])
            r, g, b = (float(v) for v in p[9:12])
            cell = atlas.crop((round(u0*AW), round(v0*AH),
                               round(u1*AW), round(v1*AH)))
            cell = cell.resize((max(1, round(w)), max(1, round(h))),
                               Image.BILINEAR)
            tint = Image.new('RGBA', cell.size,
                             (int(r*255), int(g*255), int(b*255), 255))
            alpha = cell.getchannel('A')
            tint.putalpha(alpha)
            im.alpha_composite(tint, (round(x), round(y)))
    out_path = os.path.join(ROOT, out_name)
    im.convert('RGB').save(out_path)
    print('wrote', out_path)

render('/tmp/osk2_scene1.prim', 'preview_osk_letters.png')
render('/tmp/osk2_scene2.prim', 'preview_osk_symbols.png')
render('/tmp/osk2_scene3.prim', 'preview_osk_caps.png')
render('/tmp/osk2_scene4.prim', 'preview_osk_shift.png')

File: GDS_Unity/tools/render_osk.c
/* host render harness: drives the REAL osk.c draw path with the GL seam
 * stubbed to record primitives, so a pixel-true preview of the actual C
 * layout can be rendered offline (render_osk.py rasterizes).  NOT shipped.
 * Run: gcc -O1 -o /tmp/render_osk tools/render_osk.c -lm && cd /tmp &&
 *      /tmp/render_osk && /home/user/venv/bin/python3 tools/render_osk.py */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

char gds_gamedir[1024] = "/home/user/asdf/GDS_Unity/ports/gamedevstory/gamedevstory";

static FILE *out;
int gds_egl_overlay_begin(int *sw, int *sh) { *sw = 640; *sh = 480; return 1; }
void gds_egl_overlay_rect(float x, float y, float w, float h, float r, float g, float b)
{ fprintf(out, "RECT %.2f %.2f %.2f %.2f %.3f %.3f %.3f\n", x, y, w, h, r, g, b); }
void gds_egl_overlay_rect_a(float x, float y, float w, float h, float r, float g, float b, float a)
{ fprintf(out, "RECTA %.2f %.2f %.2f %.2f %.3f %.3f %.3f %.3f\n", x, y, w, h, r, g, b, a); }
int gds_egl_overlay_atlas(const unsigned char *rgba, int w, int h)
{ (void)rgba; (void)w; (void)h; return 1; }
void gds_egl_overlay_quads(const float *v, int n, float r, float g, float b)
{
    for (int i = 0; i < n; i++) {
        const float *q = v + i * 24;
        float x = q[0], y = q[1], w = q[16] - q[0], h = q[21] - q[1];
        float u0 = q[2], v0 = q[3], u1 = q[18], v1 = q[19];
        fprintf(out, "QUAD %.2f %.2f %.2f %.2f %.5f %.5f %.5f %.5f %.3f %.3f %.3f\n",
                x, y, w, h, u0, v0, u1, v1, r, g, b);
    }
}
void gds_egl_overlay_end(void) { fprintf(out, "END\n"); }
int egl_shim_screen_w(void) { return 640; }
int egl_shim_screen_h(void) { return 480; }
long gds_mono_ms(void) { static long t = 0; t += 33; return t; }

#include "../loader_ref/osk.c"

static int find_key(int page, char lo) {
    const ovk_key_t *t = page ? g_tab_symbols : g_tab_letters;
    for (int i = 0; i < g_tab_n[page]; i++)
        if (t[i].act == OVKA_CHAR && t[i].lo == lo) return i;
    return -1;
}

int main(int argc, char **argv) {
    (void)argc; (void)argv;
    unsetenv("GDS_OSK");
    /* scene 1: letters page, "Sunny Studios" (CR-trimmed), sel on 'y' */
    out = fopen("/tmp/osk2_scene1.prim", "w");
    gds_osk_open("Company name", "Sunny Studios\r", 14);
    g_latch = 0;
    g_sel = find_key(0, 'y');
    gds_osk_draw();
    fclose(out);

    /* scene 2: flat symbols page, sel on '#', caret mid-string, maxlen
     * counter mid-FLASH (solid red), SEL cancel pill in the title band */
    out = fopen("/tmp/osk2_scene2.prim", "w");
    g_open = 0;
    gds_osk_open("Game name", "Game #1", 12);
    g_latch = 0;
    g_page = 1;
    g_sel = find_key(1, '#');
    g_shift = 0;
    g_caret = 4;                       /* caret between 'e' and ' ' */
    g_shake = 14;                      /* mid-flash: counter well into red */
    gds_osk_set_negative("Cancel");    /* cancellable prompt */
    gds_osk_draw();
    fclose(out);

    /* scene 3: caps lock on letters page, sel on Shift key: white Caps
     * cap + dark focus ring + dark badge */
    out = fopen("/tmp/osk2_scene3.prim", "w");
    g_open = 0;
    g_shake = 0;               /* keep the counter gray (scene 2 flashed) */
    gds_osk_open("Company name", "", 14);
    g_latch = 0;
    g_shift = 2;                       /* caps lock */
    for (int i = 0; i < g_tab_n[0]; i++)
        if (g_tab_letters[i].act == OVKA_SHIFT) g_sel = i;
    gds_osk_draw();
    fclose(out);

    /* scene 4: one-shot shift on the letters page: digits/punct show the
     * paired layer BIG with the base glyph small, Shift key lit gray,
     * selection ring on '7' */
    out = fopen("/tmp/osk2_scene4.prim", "w");
    g_open = 0;
    gds_osk_open("Game name", "Game", 12);
    g_latch = 0;
    g_shift = 1;
    g_sel = find_key(0, '7');
    gds_osk_draw();
    fclose(out);

    printf("scenes written, font %s\n", g_font ? "loaded" : "MISSING");
    return 0;
}
